[opentitantool] Add support for HyperDebug

jesultra opened the pull request:
Add a Transport implementation for the HyperDebug board.  The
implementation uses the rusb (Rust libusb) for finding the HyperDebug
USB device, and for bulk communication to do e.g. SPI communication.
Additionally, the impementation open the Linux /dev/ttyUSBn files
using a Rust serialport crate, which is used for serial communication
with the OPenTitan chip, as well as for sending generic text commands
to HyperDebug itself, to do e.g. GPIO-related operations.

Signed-off-by: Jes Bodi Klinke <jbk@chromium.org>

eunchan commented:
The error is:

    Commit has one or more Signed-off-by lines, but not the one we expect. We expected to find "Signed-off-by: Jes B. Klinke <jbk@chromium.org>". The commit message must contain a Signed-off-by line that matches the commit author name and email, indicating agreement to the Contributor License Agreement. See CONTRIBUTING.md for more details. You can use "git commit --signoff" to ask git to add this line for you.

cfrantz commented on sw/host/opentitanlib/src/transport/hyperdebug/spi.rs:
```diff
@@ -0,0 +1,215 @@
+        let usb_handle = Rc::clone(&hyperdebug.inner.borrow_mut().usb_handle);
+
+        // Tell HyperDebug to enable SPI bridge.
+        (*usb_handle).borrow_mut().write_control(
```
You can perform this borrow once and then use the borrowed reference throughout this function:

```
let mut handle = usb_handle.borrow_mut();

handle.write_control(...);
handle.claim_interface(...);
```

jesultra replied:
Done

cfrantz commented on sw/host/opentitanlib/src/transport/hyperdebug/spi.rs:
```diff
@@ -0,0 +1,215 @@
+    }
+}
+
+impl Drop for HyperdebugSpiTarget {
+    fn drop(&mut self) {
+        if let Err(_) = (*self.usb_handle).borrow_mut().release_interface(self.interface.interface) {
+            panic!("Error releasing SPI USB interface");
+        }
+    }
+}
```
I don't think you need to perform the release yourself.  It looks like `rusb` does this automatically:

https://docs.rs/rusb/0.9.0/src/rusb/device_handle.rs.html#114

jesultra replied:
I have dropped this section.  Worst case, the OS will release when the process terminates.

cfrantz commented on sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:
```diff
@@ -0,0 +1,388 @@
+        }
+        // Send Ctrl-C, followed by the command, then newline.  This will discard any previous
+        // partial input, before executing our command.
+        port.write(format!("\u{2}{}\n", cmd).as_bytes())?;
```
Is `\u{2}` a control-C?  (I think 2 is Ctrl-B).

cfrantz commented on sw/host/opentitanlib/src/transport/hyperdebug/spi.rs:
```diff
@@ -0,0 +1,215 @@
+            hyperdebug.spi_interface.out_endpoint, &buf[0..2], TIMEOUT)?;
+        let rc = (*usb_handle).borrow_mut().read_bulk(
+            hyperdebug.spi_interface.in_endpoint, &mut buf, TIMEOUT)?;
+        assert!(rc == 8);
+        assert!(buf[0..2] == USB_SPI_PKT_ID_RSP_USB_SPI_CONFIG.to_le_bytes());
```
I think `assert!` is going to panic if the condition is not met.  Should we use `ensure!` instead to return a more verbose error code?

jesultra replied:
Done

cfrantz commented on sw/host/opentitanlib/src/transport/hyperdebug/spi.rs:
```diff
@@ -0,0 +1,215 @@
+                    // Hyperdebug can do SPI write followed by SPI read as a single USB
+                    // request/reply.  Take advantage of that by detecting pairs of
+                    // Transfer::Write followed by Transfer::Read.
+                    assert!(wbuf.len() <= self.max_chunk_size);
+                    assert!(rbuf.len() <= self.max_chunk_size);
```
`ensure!(condition, SpiError::InvalidDataSize(...));`

https://github.com/lowRISC/opentitan/blob/master/sw/host/opentitanlib/src/io/spi.rs#L55

jesultra replied:
Done

cfrantz commented on sw/host/opentitanlib/src/transport/hyperdebug/spi.rs:
```diff
@@ -0,0 +1,215 @@
+                    self.receive(rbuf)?;
+                }
+                [Transfer::Both(wbuf, rbuf), ..] => {
+                    assert!(rbuf.len() == wbuf.len());
```
`SpiError::MismatchedDataLength`

jesultra replied:
Done

jesultra commented on sw/host/opentitanlib/src/transport/hyperdebug/gpio.rs:
```diff
@@ -0,0 +1,39 @@
+use crate::transport::hyperdebug::Hyperdebug;
+
+pub struct HyperdebugGpioPin {
+    console_tty: Box<std::path::Path>,
```
The very best in my opinion, would be if each GpioPin instance would not have its own copy, but all could refer to the same shared instance.  Maybe I can change the code to have a reference counted reference to the Inner struct of the Hyperdebug struct.  This could maybe be a fine pattern to use for several backend implementations.

jesultra commented on sw/host/opentitanlib/src/transport/hyperdebug/spi.rs:
```diff
@@ -0,0 +1,215 @@
+
+    /// Send one USB packet.
+    fn usb_write_bulk(&self, buf: &[u8]) -> Result<()> {
+        (*self.usb_handle).borrow_mut().write_bulk(self.interface.out_endpoint, buf, TIMEOUT)?;
```
You are right, claim_interface() requires a mutable borrow, but write_bulk() and read_bulk() do not.

jesultra commented on sw/host/opentitanlib/src/transport/hyperdebug/uart.rs:
```diff
@@ -0,0 +1,59 @@
+}
+
+impl HyperdebugUart {
+    pub fn open(_hyperdebug: &Hyperdebug, tty: &Box<std::path::Path>) -> Result<Self> {
```
Done

jesultra reviewed (commented):
PTAL

jesultra commented on sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:
```diff
@@ -0,0 +1,388 @@
+    }
+
+    /// Scan the USB bus for "our" devices, (shamelessly copied from
+    /// cw310, may be able to share.)
```
I have created such a util/usb.rs file, and moved common parts of HyperDebug and CW310 implementation in there.

The HyperDebug code does quite a bit of enumeration of USB interfaces/endpoints, and in this patch, some rusb structs are being returned directly from UsbBackend.  Maybe we want longer term to "translate" those various descriptors into our own structs, or maybe it is not a problem to construct rusb descriptor structs from our future mocked implementation.

cfrantz commented on sw/host/opentitanlib/src/transport/hyperdebug/uart.rs:
```diff
@@ -0,0 +1,60 @@
+impl Uart for HyperdebugUart {
+
+    fn read(&self, buf: &mut [u8]) -> Result<usize> {
+        self.port.borrow_mut().set_timeout(Duration::MAX)?;
```
I had trouble with `Duration::MAX` and serialport library:
https://github.com/lowRISC/opentitan/blob/master/sw/host/opentitanlib/src/transport/cw310/uart.rs#L20

jesultra replied:
Thanks for the warning.

cfrantz commented on sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:
```diff
@@ -0,0 +1,325 @@
+        // device, to detect minicom or another instance of
+        // opentitantool having the same serial port open.  Incoming
+        // serial data could go silenly missing, in such cases.
+        let mut buf: [u8; 128] = [0; 128];
```
This statement could be written more briefly as `let mut buf = [0u8; 128]`.

jesultra replied:
Done

cfrantz commented on sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:
```diff
@@ -0,0 +1,325 @@
+        loop {
+            match port.read(&mut buf[0..128]) {
+                Ok(rc) => {
+                    print!("Discarded {} characters: {:?}\n", rc, &std::str::from_utf8(&buf[0..rc]));
```
This should probably be a logging statement according to its severity:
```
log::info!(...);
```

stdout is reserved for machine+human readable command outputs (ie: json encoded command responses).

jesultra replied:
Done

cfrantz commented on sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:
```diff
@@ -0,0 +1,325 @@
+                            break;
+                        }
+                        _ => {
+                            print!("Error reading: {:?}\n", error);
```
`log::error!(...)`

jesultra replied:
Actually, all the information is passed into the return Err(), so I dropped this logging.

cfrantz commented on sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:
```diff
@@ -0,0 +1,325 @@
+                            if buf[i] == 10 {
+                                // Newline
+                                let mut end = i;
+                                if end > 0 && buf[end - 1] == 13 {
```
Is this protocol fundamentally line-based?  If so, would `BufReader` which has methods like `read_line`, `read_to_string` and `read_until` make this less complex?

jesultra replied:
It is line based, but I need to detect when a prompt was printed, and not terminated with a newline, which I suppose means that it would not help to use BufReader.

I have simplified the logic slightly, by removing some logic that was added in response to initial characters sometimes being lost (I later realized that the loss was caused by a minicom running in the background on the same ttyUSBx device, snubbing some characters now and then.)

cfrantz commented on sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:
```diff
@@ -0,0 +1,325 @@
+                        break
+                    }
+                }
+                Err(error) => {
```
You might be able to be more expressive with the error conditions:

```
match port.read(...) {
  Ok(rc) => { ... }
  Err(e) if e.kind == ErrorKind::TimedOut => { ... }
  Err(e) => return Err(e)
}
```

jesultra replied:
Thanks for the suggestion, doing so does save a couple of lines, and two levels of indentation.

cfrantz commented on sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:
```diff
@@ -0,0 +1,325 @@
+    #[error("Found multiple HyperDebug USB devices, use --serial")]
+    MultipleDevices,
+    #[error("Error communicating with HyperDebug")]
+    CommunicationError,
```
This particular error is used in many different places and mostly as a general error in dealing with initial setup of communications.  You might want to define this as:

```
  #[error("Error communicating with HyperDebug: {0}")]
  CommunicationError(String),
```

And then add a short string at each error site describing the nature of the error.  This will make debug easier in the event that something goes wrong with the device.

jesultra replied:
I have done so. All the strings being literals, I went for &'static str, instead of String.

cfrantz commented on sw/host/opentitanlib/src/transport/hyperdebug/spi.rs:
```diff
@@ -0,0 +1,209 @@
+        usb_handle.claim_interface(hyperdebug.spi_interface.interface)?;
+
+        // Initial bulk request/response to query capabilities.
+        let mut buf: [u8; USB_MAX_SIZE] = [0; USB_MAX_SIZE];
```
`let mut buf = [0u8; USB_MAX_SIZE];`

jesultra replied:
Done

cfrantz commented on sw/host/opentitanlib/src/transport/hyperdebug/spi.rs:
```diff
@@ -0,0 +1,209 @@
+
+        // Initial bulk request/response to query capabilities.
+        let mut buf: [u8; USB_MAX_SIZE] = [0; USB_MAX_SIZE];
+        buf[0..2].clone_from_slice(&USB_SPI_PKT_ID_CMD_GET_USB_SPI_CONFIG.to_le_bytes());
+        usb_handle.write_bulk(hyperdebug.spi_interface.out_endpoint, &buf[0..2])?;
```
I think you can write this as:

```
usb_handle.write_bulk(hyperdebug.spi_interface.out_endpoint, &USB_SPI_PKT_ID_CMD_GET_USB_SPI_CONFIG.to_le_bytes());
```
And avoid the copy into `buf`.

jesultra replied:
Done.

cfrantz commented on sw/host/opentitanlib/src/transport/hyperdebug/spi.rs:
```diff
@@ -0,0 +1,209 @@
+    /// Transmit data for a single SPI operation, using one or more USB packets.
+    fn transmit(&self, wbuf: &[u8], rbuf_len: usize) -> Result<()> {
+        let mut buf: [u8; USB_MAX_SIZE] = [0; USB_MAX_SIZE];
+        buf[0..2].clone_from_slice(&USB_SPI_PKT_ID_CMD_TRANSFER_START.to_le_bytes());
+        buf[2..4].clone_from_slice(&(wbuf.len() as u16).to_le_bytes());
+        buf[4..6].clone_from_slice(&(rbuf_len as u16).to_le_bytes());
+        let databytes = std::cmp::min(USB_MAX_SIZE - 6, wbuf.len());
+        buf[6..6 + databytes].clone_from_slice(&wbuf[0..databytes]);
```
This looks like `buf` really wants to be a struct with some named fields.  You might take a look at the `zerocopy::AsBytes` trait (used by `bootstrap/primitive.rs`) to write this as a struct and then hand it off to `usb_write_bulk` as a byte buffer.

jesultra replied:
AsBytes would take care of the sending of datagrams, but decoding received datagrams would still have to be done differently (as_bytes() on return an immutable reference).  I saw that a FromBytes trait exists, but I am not sure if it can be used without resorting to "unsafe".
Also, if we ever compile OpenTitanTool on a big-endian architecture, I am afraid that AsBytes would fail to provide the little endian encoding required in this protocol.

I would prefer to keep encoding and decoding somewhat similar.  Do you have a example of how to handle decoding of byte arrays into a struct in Rust?

cfrantz replied:
The `Frame` structures in the bootstrap implementation are already tied to a little-endian architecture.

As for writing into a struct from binary, data, I think you can derive `FromBytes` and then use `AsBytes::as_bytes_mut()` to get a mutable buffer into which you can write data.

https://docs.rs/zerocopy/0.6.1/zerocopy/trait.AsBytes.html#method.as_bytes_mut

jesultra replied:
It may be true that big-endian architectures are a thing of the past. In my first real job, our software ran on AIX, Solaris, HP Unix, and all sorts of strange machines, so keeping track of when something was in machine endian-ness and when it was in a particular wire format with fixed endian-ness was imperative.

I have changed the code to use AsBytes/FromBytes.

jesultra commented:
Hm, I did not explicitly request review from mcy@, it seems to have been a result of me pushing a new revision of the PR to github.

mcy commented on sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:
```diff
@@ -0,0 +1,320 @@
+/// Internal state of the Hyperdebug struct, this struct is reference counted such that Gpio,
+/// Spi and Uart sub-structs can all refer to this shared data, which is guaranteed to live on,
+/// even if the caller lets the outer Hyperdebug struct run out of scope.
+pub struct Inner {
```
Could you move this to just before the corresponding impl block?

jesultra replied:
Done

mcy commented on sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:
```diff
@@ -0,0 +1,320 @@
+                    len += rc;
+                    'outer: loop {
+                        for i in 0..len {
+                            if buf[i] == 10 {
```
You can write `b'\n'` for a newline byte.

jesultra replied:
Thanks. Done.

mcy commented on sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:
```diff
@@ -0,0 +1,320 @@
+        }
+        // Send Ctrl-C, followed by the command, then newline.  This will discard any previous
+        // partial input, before executing our command.
+        port.write(format!("\u{3}{}\n", cmd).as_bytes())?;
```
I would have used `\x03` here but w/e

jesultra replied:
Avoiding extra {} will probably reduce confusion. Done.

mcy commented on sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:
```diff
@@ -0,0 +1,320 @@
+                Err(error) => {
+                    match error.kind() {
+                        std::io::ErrorKind::TimedOut => {
+                            break;
```
Shouldn't rustfmt have made these into one line? e.g. `std::io::ErrorKind::TimedOut => break,`

jesultra replied:
Good point, done.

jesultra replied:
Done

mcy commented on sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:
```diff
@@ -0,0 +1,320 @@
+                    log::info!("Discarded {} characters: {:?}\n", rc,
+                               &std::str::from_utf8(&buf[0..rc]));
+                }
+                Err(error) => {
```
I think it would be more helpful to write guards into this case so as not to indent further:
```rust
Err(error) if error.kind() == TimedOut => break,
Err(error) => ...,
```

jesultra replied:
Good point, done.

mcy commented on sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:
```diff
@@ -0,0 +1,320 @@
+        // serial data could go silenly missing, in such cases.
+        let mut buf = [0u8; 128];
+        loop {
+            match port.read(&mut buf[0..128]) {
```
You should just be able to write `&mut buf` and deref coercion will do its thing.

jesultra replied:
Done

mcy commented on sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:
```diff
@@ -0,0 +1,320 @@
+        cmd: &str,
+        mut callback: impl FnMut(&str)) -> Result<()>
+    {
+        use std::io::Read;
+        use std::io::Write;
```
Can these go at the top of the file?

jesultra replied:
Done

mcy commented on sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:
```diff
@@ -0,0 +1,320 @@
+}
+
+impl Inner {
+
```
Drop this space.

mcy commented on sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:
```diff
@@ -0,0 +1,320 @@
+    fn find_tty(path: &Path) -> Result<PathBuf> {
+        for entry in fs::read_dir(path)? {
+            let dir_entry = entry?;
+            match dir_entry.file_name().into_string() {
```
This looks like it would be happier as `if let Ok(filename) = ... { ... }`

jesultra replied:
Of course, thanks.

mcy commented on sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:
```diff
@@ -0,0 +1,320 @@
+    /// tree, e.g. /sys/bus/usb/devices/1-4/1-4:1.0 .
+    fn find_tty(path: &Path) -> Result<PathBuf> {
+        for entry in fs::read_dir(path)? {
+            let dir_entry = entry?;
```
Since you're consuming `entry`, perhaps re-using the name is appropriate: `let entry = entry?;`

jesultra replied:
Done

mcy commented on sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:
```diff
@@ -0,0 +1,320 @@
+                }
+            }
+        }
+        let mut spi_names = HashMap::new();
```
@cfrantz don't we have a nice macro for creating collections?

cfrantz replied:
Yes.  In opentitanlib, we have util:collection:

You should be able to write something like:
```
use crate::util::collection;

let spi_names = collection! {
  "SPI2" => 0,
  "0" => 0,
};
```

You might need to write `let spi_names<String, u32>: = collection!{ ... }` or add `.to_string()` to the strings to force them to be the correct type.

jesultra replied:
Done

mcy commented on sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:
```diff
@@ -0,0 +1,320 @@
+        // Iterate through each USB interface, discovering e.g. supported UARTs.
+        for interface in config_desc.interfaces() {
+            for interface_desc in interface.descriptors() {
+                if let Some(idx) = interface_desc.description_string_index() {
```
You can save on indentation by writing something like
```rust
let idx = match ... {
  Some(idx) => idx,
  None => continue,
};
```

jesultra replied:
Done

mcy commented on sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:
```diff
@@ -0,0 +1,320 @@
+                            .join(format!("{}-{}:{}.{}",device.bus_number(), ports,
+                                          config_desc.number(), interface.number()));
+                        // Check the ASCII name of this USB interface.
+                        if interface_name == "HyperDebug Shell" {
```
This would probably be better as a match; you can do the UART line as `name if name.starts_with("UART") => `.

jesultra replied:
Done

mcy commented on sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:
```diff
@@ -0,0 +1,320 @@
+                            let mut in_endpoint: u8 = 0;
+                            let mut out_endpoint: u8 = 0;
+                            for endpoint_desc in interface_desc.endpoint_descriptors() {
+                                if endpoint_desc.transfer_type() == rusb::TransferType::Bulk {
```
swap this around to

```rust
if transfer_type != Bulk { continue }
```

That said, this seems like it will overwrite `in_endpoint` and such multiple times. Is that intentional?

jesultra replied:
Thanks, there is enough indentation already.  (I also removed two levels of indentation by using the trick of doing let x = match ... and then putting continue in the "unwanted" match clause.)

About the overwriting, I do rely on the USB interface having exactly one bulk endpoint in either direction, without attempting to detect if it has more than one of either.  After all, we are controlling the firmware on the USB device.

mcy commented on sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:
```diff
@@ -0,0 +1,320 @@
+                                    }
+                                }
+                            }
+                            if in_endpoint == 0 || out_endpoint == 0 {
```
Maybe `Option`s are more appropriate for this?

jesultra replied:
Done

mcy commented on sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:
```diff
@@ -0,0 +1,320 @@
+                            if in_endpoint == 0 || out_endpoint == 0 {
+                                return Err(Error::CommunicationError("Missing SPI endpoint").into());
+                            }
+                            spi_interface.replace(BulkInterface {
```
Should you handle the case where this is already filled?

jesultra replied:
Like the multiple endpoints on the same interface, which I did not try to detect. I also did not try to detect multiple interfaces both called "SPI".
It is not that hard to do after all, so I have added the if statements to do so.

mcy commented on sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:
```diff
@@ -0,0 +1,320 @@
+            match port.read(&mut buf[len..128]) {
+                Ok(rc) => {
+                    len += rc;
+                    'outer: loop {
```
Could you describe what this algorithm does? Perhaps I can figure out a more succinct way to phrase it.

jesultra replied:
The USB device has basically a console on ttyUSBn, that you could connect to, type commands, and have it do various things.  The USB device supports a very simple command editor, which includes among other things: echo of input characters, and a prompt character. This driver is connecting to that USB interface and could be sending a string like "gpioget A6\n", to which it is expecting a reply that could be "gpioget A6\n0 A6\n> ".
What code here is trying to do is first to break the input up by the newline characters, discard anything that precedes the echo of the command that it just sent, and then continue reading until it gets the two characters "> " at the beginning of a line, (and that the serial input read() times out right after that second character.)

mcy replied:
Sounds like this will do the trick:
```rust
// You may need to pull a newline off the end of cmd.
let mut split = buf.split(|b| b == b'\n').skip_while(|line| != cmd);
let _echo = split.next();
for line in split {
  if line.starts_with(b"> ") { break }
  callback(line);
}
```
That should be a starting point at a minimum.

jesultra replied:
One issue is that for performance, I want to stop after a short timeout, if the last characters received are ```"> "```, but I want to employ a much longer timeout, if the last characters are anything else.  Some commands may take a few seconds to start producing output, or maybe pause in between lines.  So I do not think that I can separate the algorithm into one part, that gathers all the data into ```buf```, and a second part which breaks ```buf``` into lines.

jesultra replied:
I have reworked the control structure, to avoid the ugly labelled outer loop, I hope this has helped make the logic easier to follow.

mcy commented on sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:
```diff
@@ -0,0 +1,320 @@
+                        break
+                    }
+                }
+                Err(error) if error.kind() == std::io::ErrorKind::TimedOut => {
```
ErrorKind is used a lot around here... `use` it?

jesultra replied:
Done.

mcy commented on sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:
```diff
@@ -0,0 +1,320 @@
+
+    // Crate SPI Target instance, or return one from a cache of previously created instances.
+    fn spi(&self, instance: &str) -> Result<Rc<dyn Target>> {
+        match self.spi_names.get(instance) {
```
`ok_or_else()` is your friend here.

jesultra replied:
Done

mcy commented on sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:
```diff
@@ -0,0 +1,320 @@
+    
+    // Crate GpioPin instance, or return one from a cache of previously created instances.
+    fn gpio_pin(&self, pinname: &str) -> Result<Rc<dyn GpioPin>> {
+        Ok(match self.inner.gpio.borrow_mut().entry(pinname.to_string()) {
```
I think that there is an `or_insert` function on Entry that is helpful here.

jesultra replied:
I could not figure out how to make that work.  The HyperdebugGpioPin::open() method is expensive, and has side effects, so we do not want to use plain or_insert(), which evaluates its "default" argument in all cases, even when the entry is populated already.  The is an or_insert_with(), which takes a closure (zero argument function) to produce the default value, but I have a Result<>, and I would like to bail out of the insertion and the call altogether, if the return of HyperdebugGpioPin::open() is Err(...), and lacking exception stack unwinding I do not know a way of doing that from inside such a closure, in Rust.

mcy replied:
Ah. yes, that is unfortunate. I don't believe Rust gives you an out here. =/

mcy commented on sw/host/opentitanlib/src/transport/hyperdebug/spi.rs:
```diff
@@ -0,0 +1,213 @@
+    
+    /// Receive data for a single SPI operation, using one or more USB packets.
+    fn receive(&self, rbuf: &mut [u8]) -> Result<()> {
+        let mut buf: [u8; USB_MAX_SIZE] = [0; USB_MAX_SIZE];
```
You shouldn't have to repeat the type; write `0u8` in the RHS

jesultra replied:
Done, here and elsewhere in this file.

mcy commented on sw/host/opentitanlib/src/transport/hyperdebug/spi.rs:
```diff
@@ -0,0 +1,213 @@
+        Ok(TransferMode::Mode0)
+    }
+    fn set_transfer_mode(&self, _mode: TransferMode) -> Result<()> {
+        unimplemented!();
```
would `todo!()` be more appropriate?

jesultra replied:
Yes, probably.  I did not know of that macro.

mcy commented on sw/host/opentitanlib/src/util/usb.rs:
```diff
@@ -0,0 +1,158 @@
+    pub fn new(
+        usb_vid: u16,
+        usb_pid: u16,
+        usb_serial: &Option<String>,
```
Wouldn't `Option<&str>` make more sense?

jesultra replied:
I believe it does. Done.

mcy commented on sw/host/opentitanlib/src/util/usb.rs:
```diff
@@ -0,0 +1,158 @@
+    timeout: Duration,
+}
+
+impl UsbBackend {
```
Most of these methods are missing comments.

jesultra replied:
My intention was that my multiline comments would cover 5 and 4 methods respectively with one comment.  Particularly the five "ugly" method that leak rusb data types, and should probably be changed in the future.  I do not want to spend too much time documenting the current behavior on each of them.

mcy replied:
Could you document your intention, then? It is very useful to try to make the output of `cargo doc` maximally friendly.

jesultra replied:
I have now documented the four "nice" methods, while leaving the five enumeration-related methods with just the one existing block comment, stating that the API they provide is ugly in that it leaks details of the rusb implementation.  In order to refine this into a proper trait, we need to find a better interface.

jesultra commented:
At first, I had missed a bunch of comments that were collapsed in my "Conversation" view.  I have now addressed all of them, and re-pushed a patchset.

mcy commented:
There's still some outstanding comments... should we keep waiting for a patch to come in?

jesultra commented:
I had almost forgotten about this PR while working on Dauntless flash protocol support.  I have made an attempt at addressing  the last open comments. PTAL.

jesultra commented:
I have rebased to lowrisc/master, and updated BUILD files (using ```glob([src/**/*.rs])```) verifying that ```bazel build //sw/host/opentitantool``` succeeds.
PTAL.

jesultra commented:
Reverted to explicitly listing .rs files in BUILD files, by request of cfrantz.
